main.nf

```diff
@@ -52,375 +61,914 @@ if (params.help) {
-    def c_white = params.monochrome_logs ? '' : "\033[0;37m"
-    def c_red = params.monochrome_logs ? '' : "\033[1;91m"
-    def c_yellow_bold = params.monochrome_logs ? '' : "\033[1;93m"
-    if (params.hostnames) {
-        def hostname = "hostname".execute().text.trim()
-        params.hostnames.each { prof, hnames ->
-            hnames.each { hname ->
-                if (hostname.contains(hname) && !workflow.profile.contains(prof)) {
-                    log.error "====================================================\n" +
-                            "  ${c_red}WARNING!${c_reset} You are running with `-profile $workflow.profile`\n" +
-                            "  but your machine hostname is ${c_white}'$hostname'${c_reset}\n" +
-                            "  ${c_yellow_bold}It's highly recommended that you use `-profile $prof${c_reset}`\n" +
-                            "============================================================"
-                }
-            }
-        }
+//----------------------------------------
+
+//----------------------------------------
+// PROCESSES: FREEBAYES
+// Variant Calling 
+// Freebayes is the default variant caller  
+//----------------------------------------
+
+process FREEBAYES {
+    tag "$sample_id"
+    publishDir "${params.outdir}/variant/freebayes",
+                pattern: "*.vcf", mode: 'copy'
+
+    input: 
+        file ref
+        tuple val(sample_id), path(bam)
+
+    output:
+        tuple val(sample_id), path(variant), path(ref), 
+                emit: 'variant'
+    
+    script: 
+        variant = "${sample_id}.vcf"
+
+    """
+    freebayes -f $ref $bam > $variant
+    """
+}
+
+//----------------------------------------
+
+//----------------------------------------
+// PROCESSES: BCFTOOLS_STATS
+// Collect stats from variant calls 
+//----------------------------------------
+
+process BCFTOOLS_STATS {
+    tag "$sample_id"
+    publishDir "${params.outdir}/variant",
+                pattern: "*_stats.txt", mode: 'copy'
+
+    input: 
+        tuple val(sample_id), path(variant), path(ref)
+
+    output:
+        file "${sample_id}_stats.txt"
+            
+    script: 
+        stats = "${sample_id}_stats.txt"
+
+    """
+    bcftools stats -F $ref $variant > $stats; 
+    
+    """
+}
+
+//----------------------------------------
+
+//----------------------------------------
+// PROCESSES: BCFTOOLS_FILTER
+// Filtering variant calls 
+//----------------------------------------
+
+process BCFTOOLS_FILTER {
+    tag "$sample_id"
+    publishDir "${params.outdir}/variant",
+                pattern: "*_filtered.vcf", mode: 'copy'
+
+    input: 
+        tuple val(sample_id), path(variant), path(ref)
+
+    output:
+        tuple val(sample_id), path(variant),
+                emit: 'variant'
+        file "${sample_id}_filtered.vcf"
+            
+    script: 
+        filtered = "${sample_id}_filtered.vcf"
+
+    """
+    bcftools stats $variant > $filtered; 
```

Comment:
This looks similar to the above `BCFTOOLS_STATS` process. Should this be `bcftools filter -i/-e "variant filtering expression"`?

Reply:
This has been changed to bcftools filter $options $variant > $filtered in the BCFTOOLS_FILTER process. $options are cmdline arguments from user

Reply:
Thanks! Could you add a default value for the filtering options? 

https://github.com/daviesdrew/variantcalling/blob/f394dfbbb69d75d33d2028ab9be2edd95ab04472/main.nf#L1127

In your `nextflow.config`, you could add something like

```nextflow
params {
  bcftools_filter_opts = " -e 'FORMAT/AO < FORMAT/RO' " // by default, exclude alternate alleles that are observed less than reference alleles
  // other params and their defaults
}
```
It looks like your params need a bit of updating as well.
https://github.com/daviesdrew/variantcalling/blob/f394dfbbb69d75d33d2028ab9be2edd95ab04472/nextflow.config#L9-L16